### ISSUE TITLE
JDR Deleted report - Notify to second user

### ISSUE THREAD

**Issue by hhovsepy** [ManageIQ/manageiq-providers-hawkular#58]
Currently if one user has deleted the JDR report which is still listed on second user's browser's Server Details page, any operation done by second user on that deleted JDR report, such as Delete or Download, will cause Unexcpected error to occur.

But it should refresh the Server's summary page with notification that selected JDR does not exist. 

Created as a separate issue based on: https://github.com/ManageIQ/manageiq-providers-hawkular/issues/56#issuecomment-331710525

**Comment by tumido** [comment]
@abonas, may I get this assigned? I'd love to continue investigation on this one.

**Comment by tumido** [comment]
The PR https://github.com/ManageIQ/manageiq-ui-classic/pull/2235 is fixing the deletion of removed JDR report part. It does deny the deletion of any report, when an exception is raised. Since I'm not able to catch or get more details on which report is causing the error, the controller just displays an error message "Unable to locate all reports in database, please try again." and refreshes the page. Thanks to that, the client receives updated list of JDR reports, so the error shouldn't happen again. There are some catches to this though:

- I'm not able to provide user the 'show' page with pre-selected reports according to his previous request. I mean, without bigger additions and changes in the front-end. So the user has to sit and click all the checkboxes again.
- Also all the actions in JDR reports are done synchronously now, which is an unnecessary load on the server. Downloading, deleting etc. should be async.
- Delete and download buttons are confusing now. `Delete` works only for selected JDR report, while `Download` doesn't care about what is selected and what is not, and downloads everything every time. The entry in the dropdown should say at least "Download ALL as text/CSV" to avoid misunderstanding. Again the action should be async.
- Download allows to get empty document, once other user deletes all reports.That shouldn't be allowed.

^ @abonas, @karelhala, @israel-hdez

**Comment by karelhala** [comment]
Personally I see couple of things that can be improved in JDR, it's usable, however some UX changes could improve it's usefulness.
* Use [miqDataTable](https://github.com/ManageIQ/ui-components/blob/master/src/gtl/components/data-table/index.ts#L4) to enable pagination, sort and option to select all records on visible page
* Use AJAX calls instead of submit. It's OK to use submit (since we do not wait for task), however combing wait for task and AJAX call, we can show spinner for whole table when delete was triggered and refresh just data inside table.
* Add some kind of JS API to select specific items, change page, sort and such to help QE team with testing

**Comment by abonas** [comment]
> Personally I see couple of things that can be improved in JDR, it's usable, however some UX changes could improve it's usefulness.
> 
> Use miqDataTable to enable pagination, sort and option to select all records on visible page
> Use AJAX calls instead of submit. It's OK to use submit (since we do not wait for task), however combing wait for task and AJAX call, we can show spinner for whole table when delete was triggered and refresh just data inside table.
> Add some kind of JS API to select specific items, change page, sort and such to help QE team with testing
> So mostly 2 points are the biggest improvement, use AJAX calls, and some kind of prettier table with pagination and such to allow users easier work with it.

@karelhala could you please open this as a separate issue and also issue the relevant fixes? :)

**Comment by israel-hdez** [comment]
@tumido I cannot see a way to make _download_ actions async. You need to pass the file to the browser for download, and making this async won't allow it. Or maybe I am missing a clever strategy here?

Delete action for sure can be made async, and if combined with @karelhala suggestions, it sounds great.

**Comment by tumido** [comment]
@israel-hdez the approach I would choose, is to use a `initiate_wait_for_task` and follow the implementation as for cloud controllers. Once the task is finished just serve the file to/from UI. Downloading many JDRs causes a query to DB for each report, which shouldn't be done: 1st in a blocking manner and 2nd in the UI worker at all. Also the same for the deletion. Also we have to keep in mind that the UI worker can live in a completely different environment which may decrease performance a lot once the UI worker is pooling for each report.

TL;DR: For the downloading, I would suggest to prepare the file as a task placed in the queue worker, and once the task is finished, serve the file in UI worker via some `javascript_redirect` or such...

**Comment by israel-hdez** [comment]
@tumido are you talking about the "download CSV/Text" or the "Download" button to get the JDR zip file?
If it's the later, for me it doesn't make sense to make it async.

If it's the former, it makes _some_ sense. From a technical point of view, it will be nice if "Download CSV/Text" is async. From a performance point of view, you are just moving the process from one worker to another one, and I think it will add more overhead mainly because the JDRs list most likely won't be large and also because it will hit `MiqQueue` which is not great right now.

**Comment by tumido** [comment]
@israel-hdez I'm talking about the first one, the  "Download as CSV/Text". There it make sense to me, I'm not sure about the later one (why not?).

What do you mean by _some_ sense? Because to me the behavior when the UI is basically stuck until all JDRs are fetched from DB and merged into one file, is not what I would, as a user, expect. Also I don't understand (lack of knowledge on my side), what's wrong with `MiqQueue`. And why should we avoid using it for tasks which are done on the server/DB side anyways? I mean, when there's one JDR report, sure, that's simple and consts one request to DB. But when there's tons of JDRs the UI worker shouldn't be doing it. What piece I'm missing here? I'm sorry but I still think it make sense what I'm saying here, I'd like to be proven wrong. ;)

I'm maybe too daring but I would like to hear what @Ladas thinks about this...

**Comment by israel-hdez** [comment]
@tumido Apparently, we have started a fight! :sweat_smile: 

Hmm... I want to ensure that this is clear: when you use "Download as CSV/Text" you are not downloading the JDRs. Instead, you are downloading the table that you are seeing in the UI, but in CSV or Text format. So, the JDRs aren't being merged and thus the UI isn't reading the lots bytes of the JDRs. It's just the listing.

The problem with `MiqQueue` is that is has performance issues and the core team is working on figuring out a solution. A task coming into the queue may remain a long time there until it's picked for processing. So, if the task/process is simple and fast, I wouldn't use it.

IMO, the "Download as CSV/Text" is simple enough to avoid `MiqQueue` and let the UI process it. But if you think it will have better performance by making it async, I think you should do "the research" and bring some numbers with the timings for the asynchronous and the synchronous cases. There are lots of places that have a "Download as CSV/Test" button working synchronously and if your suggestion has more performance, your code should be in Miq-core to be widely available :)

**Comment by tumido** [comment]
@israel-hdez no fight, please 😄 I'm still totally new to project so let's consider it rather an "educating @tumido on the matter" than "fight". 😄

I see. That changes the whole meaning to me. That was probably the missing piece in puzzle! Than both approaches may be fine to me - queue it or do it on the UI worker. Yet I still think, it might be better to do it async. For example, on my setup, when there's more JDRs, it blocks the UI for couple of seconds. It can also be possible to leave it as a sync task and show some spinner (like when you're logging in etc.). Honestly, for me, it's not much about the total time spent by generating the report, rather UX. If i can choose between having the client responsive rather than be blocked by open request (waiting for UI worker to get all JDRs), I would prefer the first option here.

In other words, as a user, I don't mind waiting for the report when I'm able to work with ManageIQ and do other tasks.

If you think we're wasting our time discussing it here (I have a suspicion it might be that case), please ping me and we can stop bothering everybody else. :)

**Comment by israel-hdez** [comment]
@tumido Let's take advantage that we had our discussion here to ask feedback from @abonas and @karelhala. If they agree with you, you can have green light to do the enhancement. Then, we could continue out of the thread if needed :)

**Comment by abonas** [comment]
@israel-hdez @tumido the "download as csv..." is indeed not downloading the jdr report content, but rather the table. That functionality is similar to any "download as csv..." other tables have in miq UI, so if anything should be considered for redesign, I see this is a core feature and not a middleware one. Specifically here, I don't see a lot of burden on user experience if a user has to wait a few seconds (and downloading this is not a task done often) soI do not think it is worth the extra effort. @tumido if you still think this is worth to be redesigned in all miq, it's best to discuss with miq UI core team.

**Comment by Ladas** [comment]
@karelhala @tumido @israel-hdez the general rule of thumb is to do async job anytime we talk to the provider or do long running job.

Main Reasons:
1. We support architectures where UI workers can't see the providers. In that case we have a specific appliances with backend workers, that are routed to provider and none other can see it or talk to the provider (with web consoles being routed via proxies as an exception).
2. Not to block UI workers with potential long running job, cause that will prevent anybody from accessing the UI. (e.g. just api call timeout can nicely block the whole UI)

The MiqQueue problems are being handled, in the meantime we still follow this rule of thumb everywhere. Especially because of the point 1., where all sync features would stop working if we wouldn't follow it.

**Comment by abonas** [comment]
@Ladas have you seen [this comment](https://github.com/ManageIQ/manageiq-providers-hawkular/issues/58#issuecomment-332814172)?
if there is anything that requires a change in "download csv/text", it is a core functionality.

**Comment by Ladas** [comment]
@abonas if it's just touching MIQ DB and returning the CSV base on that table, then yeah the core would have to be redesigned (at best we would have a separate object storage service, where we would store it while providing link)

**Comment by hhovsepy** [comment]
Tested on master.20171005031204_874cad6

Issue is fixed partially. For deleting reports it's fine.
However it still fails for Download of deleted reports.
![screenshot from 2017-10-05 10 05 06](https://user-images.githubusercontent.com/604313/31216799-232d4df6-a9b5-11e7-8d14-0238b0e56deb.png)

**Comment by tumido** [comment]
@hhovsepy, this should be completely redesigned by @karelhala, I'll sync with him and take a look if the issue is still present with his implementation. Thanks for catching that!

**Comment by tumido** [comment]
@hhovsepy so, I can confirm that this issue is still present and I have prepared a patch for it. However, I'm not able to test it. Can anyone help me with that?

@israel-hdez, how did you managed to prepare a environment where the WildFly is actually generating JDRs and such? I'm not able to do so via hawkinit, what am I doing wrong?

@hhovsepy, do you have any Hawkular instance available which I can borrow for a moment, plug it into my ManageIQ and debug it?

**Comment by israel-hdez** [comment]
@tumido what error are you getting when trying to generate the JDRs? A Hawkinit environment should work fine if you are using all _latest_ versions (except for cassandra).

**Comment by tumido** [comment]
@israel-hdez lol, it does, I've figured that out just before reading your comment. My belief in released and RC versions suffers now. :P

**Comment by hhovsepy** [comment]
Verified on master.20171017060733_05c1495 
Download of deleted report does not crash.
